libnmstate/nm/wired.py

```diff
@@ -44,12 +44,17 @@ def create_setting(iface_state, base_con_profile):
 
     if auto_negotiation:
         wired_setting.props.auto_negotiate = True
-        if not speed:
+        if not speed and not duplex:
```

Comment:
I think we should not set here the wired setting, only update the speed per what is needed.
Then, at the stage below, we apply everything in once location.
The function is becoming too long, please also consider splitting it into smaller functions.

I'm also not clear why you merged speed & duplex cases.

Reply:
If speed or duplex are provided by the user, both need to be set for Network Manager. The missing value needs to be taken from the current device state.

Setting the speed later could work, for duplex we would need a special value that makes it possible to distinguish it from not being set by the user since it is set to None.

Reply:
When we get here, all 3 values exist, either as None, as a user input (+ "sanitized" the input) or as a merge result.
My question is, why can't we handle them individually:
if speed is None:
   speed = ...
if duplex is None:
   duplex = ...

Then, apply explicitly.

And now that I think about it, why do we need to read the speed/duplex? The merge from the netinfo should do that, no?

Reply:
There was a bug in the previous code, therefore I had to change it a little bit.

auto-negotiation, speed and duplex are set to None if they are not specified by the user and therefore are not merged in netinfo (this is what the sanitization step does) and we need to get them from the device if the values are needed.

When only auto-negotiation was specified and set to True, then nmstate needs to reset the speed and duplex settings to their default values (`0` and `None`). When we first set `speed` to `None` in this case and later want to apply this value to `props.duplex`, we cannot distinguish "duplex was not specified by the user" (in which case it would be `None`) from "duplex needs to be set to None  for auto-negotiation". Therefore I set the props directly when it is clear that it needs to be set to None.

For `speed` it could work if I change the conditional to `if speed is not None` since `0` is false.

Reply:
Maybe I'm missing something, but why do we care at the "apply" stage the original source of why `duplex` is `None`?
After the desired state has been processed: normalized/sanitized/canonicalized/metadatalized , we no longer should care why it is like it is, the lower "drivers" should just apply it.

What am I missing here?

Reply:
Uh, somehow my comment got lost. :-/

I guess the confusion comes from where the nmstate state is converted into network manager config. NM allows more than nmstate, for example NM allows to define a "use the default speed setting" but nmstate only supports actual values. At the same time, nmstate allows to only set partial aspects of the state (for example disabling auto-negotiation without setting speed or duplex).

Somewhere in nmstate, we need to be able to decide what to do in which case. Currently this is happening in the NM create setting code for the wired setting and speed/duplex are passed as None, if the user did not specify them. However, when auto-negotiation is disabled and the user does not specify speed/duplex, then IMHO nmstate should use the current runtime speed/duplex settings. Therefore the NM create setting code needs to fetch them, since they are not in the function variables (they are set to None).

Reply:
> when auto-negotiation is disabled and the user does not specify speed/duplex, then IMHO nmstate should use the current runtime speed/duplex settings.

It makes sense to me too, I would just argue that this logic fits the "merging" step and not in the final step that applies the state. When looking at the function now, it is clearly doing more than one thing (at its abstract level), which could be avoided if we move this logic away.

I would have preferred smaller functions as well, IMO it is a bit messy at the moment.

Lets try to refactor this later on when we'll have functional/integration tests in place.